### PR TITLE
Allow simultaneous rebasing of duplicate commits, duplicating multiple commits, undo after duplicate

### PR DESCRIPTION
**Pull request by ilyagr**
Fixes https://github.com/martinvonz/jj/issues/27, https://github.com/martinvonz/jj/issues/694 and https://github.com/martinvonz/jj/issues/1050

I combined the bugfix PRs with the feature PR into one since they are both ready. If a portion of this PR gets a faster review than the rest, I'll split them again.

Now also includes a minor fix to `jj abandon`: if the same revision is specified several times, it is now only listed once in the output.

# Checklist

If applicable:
- [x] I have updated `CHANGELOG.md`
- [x] Updated command-line help
- [x] I have added tests to cover my changes

### CODE REVIEW

**Comment by ilyagr** [comment]
I made this a draft until I figure out whether we can fix https://github.com/martinvonz/jj/issues/1050 or I should work around it by not using `undo` in tests.

File: lib/src/store.rs [hunk 1]
```diff
@@ -81,11 +81,24 @@ impl Store {
...
     }
-
     pub fn write_commit(self: &Arc<Self>, commit: backend::Commit) -> BackendResult<Commit> {
         assert!(!commit.parents.is_empty());
-        let commit_id = self.backend.write_commit(&commit)?;
-        let data = Arc::new(commit);
+        let mut mut_commit = commit;
+        let commit_id = loop {
+            // It's possible that another commit with the same commit id (but different
+            // change id) already exists. Fundge the timestamp until this is no
```
Comment (samueltardieu) [inline lib/src/store.rs:89]: Typo: Fundge -> Fudge (or "Slightly decrease")

Comment (ilyagr) [reply lib/src/store.rs:89]: Thank you! I will un-fundge it.

File: tests/test_duplicate_command.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    o 04db325bf42c test-username@host.example.com 2001-02-03 04:05:09.000 +07:00 - 2001-02-03 04:05:09.000 +07:00
+    | commit working copy
+    o 682c21f25d34 test-username@host.example.com 2001-02-03 04:05:08.000 +07:00 - 2001-02-03 04:05:08.000 +07:00
+    | new empty commit
+    | args: jj new root -m a
+    o a99a3fd5c51e test-username@host.example.com 2001-02-03 04:05:07.000 +07:00 - 2001-02-03 04:05:07.000 +07:00
+    | add workspace 'default'
+    o 56b94dfc38e7 test-username@host.example.com 2001-02-03 04:05:07.000 +07:00 - 2001-02-03 04:05:07.000 +07:00
+      initialize repo
+    "###);
```
Comment (martinvonz) [inline tests/test_duplicate_command.rs:269]: Do we really need the `jj op log` output to confirm that the undo worked? The regular log output that you have at the end seems enough.

File: lib/src/store.rs [hunk 3]
```diff
@@ -81,11 +81,24 @@ impl Store {
...
     pub fn write_commit(self: &Arc<Self>, commit: backend::Commit) -> BackendResult<Commit> {
         assert!(!commit.parents.is_empty());
-        let commit_id = self.backend.write_commit(&commit)?;
-        let data = Arc::new(commit);
+        let mut mut_commit = commit;
+        let commit_id = loop {
+            // It's possible that another commit with the same commit id (but different
+            // change id) already exists. Adjust the timestamp until this is no
+            // longer the case.
+            let maybe_commit_id = self.backend.write_commit(&mut_commit);
```
Comment (martinvonz) [inline lib/src/store.rs:91]: Can we move the retry into the Git backend? Then we shouldn't have to add the `AlreadyHaveCommitWithId` variant. The error is specific to the Git backend, so it would be nice to not have it pollute the API.

File: src/commands.rs [hunk 4]
```diff
@@ -2028,30 +2029,91 @@ fn cmd_commit(ui: &mut Ui, command: &CommandHelper, args: &CommitArgs) -> Result
     Ok(())
 }
 
+fn resolve_multiple_rewriteable_revsets(
+    revisions: &[RevisionArg],
```
Comment (martinvonz) [inline src/commands.rs:2033]: nit: `revision_args` is probably clearer, especially since we have another `revisions` variable inside the loop

File: tests/test_duplicate_command.rs [hunk 5]
```diff
@@ -89,6 +89,144 @@ fn test_duplicate() {
...
+    | o | c0cb3a0b73e7   c
+    |/ /  
+    | o 1394f625cbbd   b
+    |/  
+    o 2443ea76b0b1   a
+    o 000000000000   (no description set)
+    "###);
+
+    test_env.jj_cmd_success(&repo_path, &["undo"]);
+    let stdout = test_env.jj_cmd_success(&repo_path, &["duplicate", "b:", "d"]);
```
Comment (martinvonz) [inline tests/test_duplicate_command.rs:138]: We should also have a test of `jj duplicate b b` or `jj duplicate b: d:` or similar, i.e. where some revisions are mentioned multiple times.

File: CHANGELOG.md [hunk 6]
```diff
@@ -77,7 +77,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Fixed handling of escaped characters in .gitignore (only keep trailing spaces
   if escaped properly).
 
-* `jj undo` now works after `jj duplicate`.
+* `jj duplicate` followed by `jj rebase` of a tree containing both the original
+  and duplicate commit no longer crashes. Also, `jj undo` now works after
+  `jj duplicate`.
```
Comment (martinvonz) [inline CHANGELOG.md:86]: The fix is general enough to fix #27, right? If so, add that to the commit message (or PR), and make the this changelog entry more generic.

Hooray, btw! :)

**Review by martinvonz — COMMENTED** [review]
I didn't finish reviewing the last two commits. I'll probably get around to that later tonight.

File: lib/src/store.rs (hunk 3 above)
Comment (ilyagr) [reply lib/src/store.rs:91]: I tried that, and it turned into a mess. The git backend locks the repo for writing, and doing it naively makes it try to lock an already-locked `mut_repo` and hang forever. 

If I had to do it, I'd rename GitBackend::write_commit into `write_commit_internal` and wrap it into another function. Does that sound good to you?

File: tests/test_duplicate_command.rs (hunk 2 above)
Comment (ilyagr) [reply tests/test_duplicate_command.rs:269]: No, it was left over from debugging. I'll get rid of it.

File: CHANGELOG.md (hunk 6 above)
Comment (ilyagr) [reply CHANGELOG.md:86]: I don't understand that bug very well. I'm sure you are right, will look into it and put it here. 🎉

File: lib/src/store.rs (hunk 3 above)
Comment (martinvonz) [reply lib/src/store.rs:91]: Sounds good. Thanks

File: CHANGELOG.md (hunk 6 above)
Comment (ilyagr) [reply CHANGELOG.md]: I think that's about right. I'm not sure if there are any instances of #27 left apart from #694 but, if so, they should be resolved.

File: lib/src/git_backend.rs [hunk 7]
```diff
@@ -445,6 +445,28 @@ impl Backend for GitBackend {
     }
 
     fn write_commit(&self, contents: &Commit) -> BackendResult<CommitId> {
+        let mut mut_commit = contents.clone();
+        // It's possible a commit already exists with the same commit id but different
+        // change id. Adjust the timestamp until this is no longer the case.
+        loop {
+            match self.write_commit_internal(&mut_commit)? {
```
Comment (martinvonz) [inline lib/src/git_backend.rs:452]: Much of `write_commit_internal()` could be moved out of it and into this function (we shouldn't have to write the tree on every iteration, for example), but I don't think it's worth it since this happens so rarely.

File: lib/src/git_backend.rs [hunk 8]
```diff
@@ -756,13 +775,8 @@ mod tests {
...
-        match store.write_commit(&commit2) {
-            Ok(_) => {
-                panic!("expectedly successfully wrote two commits with the same git commit object")
-            }
-            Err(BackendError::Other(message)) if message.contains(&expected_error_message) => {}
-            Err(err) => panic!("unexpected error: {err:?}"),
-        };
+        // `write_commit` should prevent the ids from being the same by changing the
+        // committer timestamp of the commit it actually writes.
+        assert_ne!(store.write_commit(&commit2).unwrap(), commit_id1);
```
Comment (ilyagr) [inline lib/src/git_backend.rs:780]: It's curious that the underflow caused by subtracting a second from the 0 timestamp doesn't cause any problem here. I expected Rust to panic.

I think it's safe to assume the real timestamp is far from 0, but we could check and panic explicitly if we wanted.

I could also change the timestamp in the test, but I found it interesting that the code works even in this case.

File: src/commands.rs [hunk 9]
```diff
@@ -2028,30 +2029,91 @@ fn cmd_commit(ui: &mut Ui, command: &CommandHelper, args: &CommitArgs) -> Result
...
+        for commit in to_duplicate.iter() {
+            if !commit
+                .parents()
+                .iter()
+                .any(|parent| to_duplicate.contains(parent))
+            {
+                original_commit = Some(commit.clone());
+                break;
+            }
+        }
```
Comment (martinvonz) [inline src/commands.rs]: You should be able to call `repo.index().topo_order()` to get a topological order. I don't think that de-duplicates commits, so you'll still have to take care of that here (skipping commits already in `duplicated_old_to_new`).

File: src/commands.rs [hunk 10]
```diff
@@ -2062,18 +2124,7 @@ fn cmd_abandon(
...
-            let revisions = workspace_command.resolve_revset(revset)?;
-            workspace_command.check_non_empty(&revisions)?;
-            for commit in &revisions {
-                workspace_command.check_rewriteable(commit)?;
-            }
-            acc.extend(revisions);
-        }
-        acc
-    };
+    let to_abandon = resolve_multiple_rewriteable_revsets(&args.revisions, &workspace_command)?;
```
Comment (martinvonz) [inline src/commands.rs:2136]: Not for this PR, but it seems that this would also need to be de-duplicated, so `jj abandon @ @` doesn't print the commit twice

File: lib/src/git_backend.rs (hunk 7 above)
Comment (martinvonz) [reply lib/src/git_backend.rs:452]: That was obviously supposed to be a reply to your comment about underflow when subtracting from the timestamp

File: src/commands.rs [hunk 11]
```diff
@@ -2035,6 +2035,7 @@ fn cmd_duplicate(
 ) -> Result<(), CommandError> {
     let mut workspace_command = command.workspace_helper(ui)?;
     let predecessor = workspace_command.resolve_single_rev(&args.revision)?;
+    workspace_command.check_rewriteable(&predecessor)?;
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply src/commands.rs]: At some point, I renamed `check_rewriteable` to `check_not_root_commit`, but then I decided that there isn't much point. As you said, changing the message didn't seem worth it.

File: src/commands.rs (hunk 10 above)
Comment (ilyagr) [reply src/commands.rs:2136]: Now that we have `IndexSet`, it's trivial to do and won't change the order of things.

File: lib/src/git_backend.rs (hunk 7 above)
Comment (ilyagr) [reply lib/src/git_backend.rs:452]: > Much of write_commit_internal() could be moved out of it and into this function

I don't see many advantages of doing that. I think that if we tried really hard, we could save a single clone of a Commit (which should be quite fast). 

> We actually allow negative timestamps

That's neat. But why? Is someone somewhere working on turning the process of writing the Gutenberg Bible into a git repository?

Comment (martinvonz) [reply lib/src/git_backend.rs:452]: > > Much of write_commit_internal() could be moved out of it and into this function
> 
> I don't see many advantages of doing that. I think that if we tried really hard, we could save a single clone of a Commit (which should be quite fast).

We would avoid writing the tree object multiple times, right? Again, I think it's rare enough that the retries happen that we shouldn't bother.

> 
> > We actually allow negative timestamps
> 
> That's neat. But why? Is someone somewhere working on turning the process of writing the Gutenberg Bible into a git repository?

The link in my commit message is stale. https://github.com/git/git.github.io/blob/master/_posts/2022-09-31-edition-91.markdown is where it's now. The example mentioned there was "Apollo code".

Comment (ilyagr) [reply lib/src/git_backend.rs:452]: There's a joke there somewhere about rebasing Apollo code on top of the Space Shuttle code to get the code for the SLS.

Comment (martinvonz) [reply lib/src/git_backend.rs:452]: > We would avoid writing the tree object multiple times, right?

I misread the code; we don't write trees in `write_commit()`. We could avoid *reading* the tree, though.

File: src/commands.rs (hunk 10 above)
Comment (ilyagr) [reply src/commands.rs:2120]: I added this in. As I hoped, all I needed to do was to replace a `Vec` with an `IndexSet`. This meant that I also moved the commit with the other test for `jj abandon` towards the end of the PR.

File: src/commands.rs (hunk 9 above)
Comment (ilyagr) [reply src/commands.rs]: Done. This shortened things a bit.

**Review by martinvonz — APPROVED** [review]

**Comment by ilyagr** [comment]
Thank you for the review and the help! :)
